subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java

```diff
@@ -16,15 +16,22 @@
 
 package org.gradle.nativeplatform.toolchain.internal.metadata;
 
+import org.gradle.api.Action;
 import org.gradle.platform.base.internal.toolchain.SearchResult;
 
 import java.io.File;
 import java.util.List;
 
 public interface CompilerMetaDataProvider<T extends CompilerMetadata> {
 
-    SearchResult<T> getCompilerMetaData(File binary, List<String> additionalArgs, List<File> searchPath);
+    SearchResult<T> getCompilerMetaData(List<File> searchPath, Action<? super CompilerExecSpec> configureAction);
```

Comment:
I'm not quite sure I understand why this is necessary.  Does DEVELOPER_DIR need to be set in order to run the compiler to get information about it?

Reply:
The `man xcode-select` have a section that mentions every tool in the following list depends on the `DEVELOPER_DIR` (or `xcode-select`) to switch version:
```
FILES
       /usr/bin/xcrun
          Used to find or run arbitrary commands from the active developer directory. See xcrun(1) for more information.

       /usr/bin/actool  /usr/bin/agvtool  /usr/bin/desdp  /usr/bin/genstrings  /usr/bin/ibtool /usr/bin/ictool /usr/bin/instruments /usr/bin/iprofiler /usr/bin/opendiff /usr/bin/sdef /usr/bin/sdp /usr/bin/stapler
       /usr/bin/xcodebuild /usr/bin/xed
          Runs the matching Xcode tool from with the active developer directory.

       /usr/bin/BuildStrings /usr/bin/CpMac /usr/bin/DeRez /usr/bin/GetFileInfo /usr/bin/MergePef /usr/bin/MvMac /usr/bin/ResMerger /usr/bin/RezDet /usr/bin/RezWack /usr/bin/Rez  /usr/bin/SetFile  /usr/bin/Split-
       Forks   /usr/bin/UnRezWack  /usr/bin/ar  /usr/bin/as  /usr/bin/asa  /usr/bin/bison  /usr/bin/c89  /usr/bin/c99  /usr/bin/clang++  /usr/bin/clang  /usr/bin/cmpdylib  /usr/bin/codesign_allocate  /usr/bin/cpp
       /usr/bin/ctags  /usr/bin/ctf_insert  /usr/bin/dsymutil  /usr/bin/dwarfdump  /usr/bin/flex++   /usr/bin/flex   /usr/bin/g++   /usr/bin/gatherheaderdoc   /usr/bin/gcc   /usr/bin/gcov   /usr/bin/git-cvsserver
       /usr/bin/git-receive-pack  /usr/bin/git-shell  /usr/bin/git-upload-archive  /usr/bin/git-upload-pack  /usr/bin/git /usr/bin/gm4 /usr/bin/gnumake /usr/bin/gperf /usr/bin/hdxml2manxml /usr/bin/headerdoc2html
       /usr/bin/indent /usr/bin/install_name_tool /usr/bin/ld /usr/bin/lex  /usr/bin/libtool  /usr/bin/lipo  /usr/bin/lldb  /usr/bin/lorder  /usr/bin/m4  /usr/bin/make  /usr/bin/mig  /usr/bin/mkdep  /usr/bin/nasm
       /usr/bin/ndisasm   /usr/bin/nm   /usr/bin/nmedit   /usr/bin/objdump   /usr/bin/otool   /usr/bin/pagestuff  /usr/bin/ranlib  /usr/bin/rebase  /usr/bin/redo_prebinding  /usr/bin/resolveLinks  /usr/bin/rpcgen
       /usr/bin/segedit /usr/bin/size /usr/bin/strings /usr/bin/strip /usr/bin/svn /usr/bin/svnadmin /usr/bin/svndumpfilter /usr/bin/svnlook /usr/bin/svnserve /usr/bin/svnsync  /usr/bin/svnversion  /usr/bin/swift
       /usr/bin/swiftc /usr/bin/unifdef /usr/bin/unifdefall /usr/bin/xml2man /usr/bin/yacc
          Runs the matching BSD tool from with the active developer directory.
```

The tools are in a single location but redirect to the right Xcode installation. We can't use `xcode-select` as concurrent execution would cause the flakiness. To query the right metadata, the compiler needs to be invoked using the right `DEVELOPER_DIR` variable. The compiler (the official way from my understanding) is always located in the same directory so we can't infer the right value from the compiler path. The changes have effects only for our testing infrastructure at the moment. It also has the side-effect of switching compiler version if someone starts Gradle with that environment variable set. I kept it internal and dependent on the environment variable for now as we supporting a public API needs more work. In the future, we could see the tool chain discovery able to manager Xcode version properly. Right now, if you only set the path to the compiler, you may still get incorrect result as `xcrun` and co. will assume the default Xcode (from `/Applications/Xcode.app`).

I assumed the worker API wouldn't carry the `DEVELOPPER_DIR` if it fork the work to another process so I pass the value were it's required.